# MATRIKS INVERS

adalah kebalikan dari suatu matriks. jadi jika dilambangkan suatu matriks adalah matriks A
 maka kebalikan atau invers dari matriks A
 adalah matriks A-1
. Matriks A-1
 dapat dikatakan invers dari matriks A
 jika memenuhi syarat apabila matriks A
 dikalikan dengan matriks A-1
 menghasilkan matriks I
 yang mana matriks I
 adalah matriks identitas.

## Sifat sifat Matriks Invers
* Invers dari matriks Invers adalah matriks itu sendiri, contoh invers dari matrik adalah matriks itu sendiri
* Perkalian matriks dengan inversnya akan menghasilkan matriks identitas, contoh
* Invers dari perkalian dua matriks adalah hasil dari perkalian inversnya dalam urutan terbalik, contoh
* Invers dari matriks transpose adalah transpose dari matriks invers, contoh
* Invers dari matriks identitas adalah matriks identitas itu sendiri, contoh

# Eliminasi Gauss Jordan
Eliminasi Gauss adalah suatu cara mengoperasikan nilai-nilai di dalam matriks sehingga menjadi matriks yang lebih sederhana. Metode Eliminasi Gauss adalah salah satu cara yang paling awal dan banyak digunakan dalam penyelesaian sistem persamaan linier. Cara ini ditemukan oleh Carl Friedrich Gauss. Prosedur penyelesaian dari metode ini adalah dengan melakukan operasi baris sehingga matriks tersebut menjadi matriks yang Eselon-baris. Ini dapat digunakan sebagai salah satu metode penyelesaian persamaan linear dengan menggunakan matriks. Caranya dengan mengubah persamaan linear tersebut ke dalam matriks teraugmentasi dan mengoperasikannya. Setelah menjadi matriks Eselon-baris, lakukan substitusi balik untuk mendapatkan nilai dari variabel-variabel terseb

# Ciri-ciri eliminasi gaus jordan
* Jika suatu baris tidak semua nol, maka bilangan pertama yang tidak nol ada 1(1 utama).
* Baris nol terletak paling bawah.
* 1 utama baris berikutnya berada dikanan 1 utama baris diatasnya.
* Dibawah 1 utama harus nol.



## Mencari invers dengan eliminasi gaus jordan

Dalam penerapan gauss Jordan dengan menggunakan Operasi Baris Elementer atau disebut OBE. Operai OBE terdiri dari 3 cara, yaitu:

1. Menambahkan kelipatan baris ke baris yang lain

2. mengkalikan baris dengan bilangan selain nol

3. Menukarkan posisi baris satu dengan baris yang lain

Untuk mencari Invers dari sebuah Matrix kita harus memperhatikan persamaan - persamaannya. Untuk bisa memasukkannya dalam bentuk matrix kita pisah antara koefisien yang ada disebelah ruas kiri dan konstanta yang ada disebelah ruas kanan. contoh saya mempunyai persamaan linier 4 variabel:

$$
0X_1 + X_2 + 0X_3 + 1X_4  = 5 \\

X_1 + -3X_2 + 0X_3 + 2X_4  = 8 \\

1X_1 + -1X_2 + 3X_3 + X_4  = 6 \\

2X_1 + 4X_2 + 0X_3 + 5X_4  = 7 \\
$$

Lalu kita akan masukkan dalam bentuk matrix:

$$
\begin{align} A= \begin{bmatrix} 0&1&0&1\\ 0&-3&0&2\\ 1&-1&3&1\\ 2&4&0&5\\  \end{bmatrix} x= \begin{bmatrix} x_1\\ x_2\\ x_3\\ x_4\\ \end{bmatrix} b= \begin{bmatrix} 5\\ 8\\ 6\\ 7\\ \end{bmatrix} \end{align}
$$

$$
\begin{align} I= \begin{bmatrix} 1&0&0&0\\ 0&1&0&0\\ 0&0&1&0\\ 0&0&0&1\\  \end{bmatrix} \end{align}
$$

Setelah itu, kita ubah menjadi augmented matrix yang ditambahkan dengan matriks identitas. Kita lambangkan sebagai matriks AI

$$
\begin{align} AI= \begin{bmatrix} 0&1&0&1&1&0&0&0\\ 0&-3&0&2&0&1&0&0\\ 1&-1&3&1&0&0&1&0\\ 2&4&0&5&0&0&0&1\\  \end{bmatrix} \end{align}
$$

## Penyelesaian dengan code

Berikut adalah contoh penyelesaian mencari Invers Matrix 4x4 dengan code python

In [1]:
# import library numpy
import numpy as np

def RowSwap(A,k,l):
# =============================================================================
#     A adalah sebuah NumPy array.  RowSwap akan menghasilkan duplicate dari
#     array dengan baris k dan l di tukar
# =============================================================================
    m = A.shape[0]  # m adalah nomor baris di A
    n = A.shape[1]  # n adalah nomor kolom di A

    B = np.copy(A).astype('float64')

    for j in range(n):
        temp = B[k][j]
        B[k][j] = B[l][j]
        B[l][j] = temp

    return B

def RowScale(A,k,scale):
# =============================================================================
#     A adalah sebuah NumPy array.  RowScale akan menghasilkan duplicate dari
#     array dengan baris k di kali dengan skalar bukan 0
# =============================================================================
    m = A.shape[0]  # m adalah nomor baris di A
    n = A.shape[1]  # n adalah nomor kolom di A

    B = np.copy(A).astype('float64')

    for j in range(n):
        B[k][j] *= scale

    return B

def RowAdd(A,k,l,scale):
# =============================================================================
#     A adalah sebuah NumPy array.  RowAdd akan menghasilkan duplicate dari
#     array dengan baris k akan di kali dengan 'scale' bukan 0. lalu nilai
#     baris l akan di tambah dengan nilai baris k yang sudah dikalikan
# =============================================================================
    m = A.shape[0]  # m adalah nomor baris di A
    n = A.shape[1]  # n adalah nomor kolom di A

    B = np.copy(A).astype('float64')

    for j in range(n):
        B[l][j] += B[k][j]*scale

    return B

In [2]:
BI = np.array([[0,1,0,1,1,0,0,0],[0,-3,0,2,0,1,0,0],[1,-1,3,1,0,0,1,0],[2,4,0,5,0,0,0,1]])
print(BI)

[[ 0  1  0  1  1  0  0  0]
 [ 0 -3  0  2  0  1  0  0]
 [ 1 -1  3  1  0  0  1  0]
 [ 2  4  0  5  0  0  0  1]]


In [3]:
BI1 = RowAdd(BI,0,1,4)
print(BI1)

[[ 0.  1.  0.  1.  1.  0.  0.  0.]
 [ 0.  1.  0.  6.  4.  1.  0.  0.]
 [ 1. -1.  3.  1.  0.  0.  1.  0.]
 [ 2.  4.  0.  5.  0.  0.  0.  1.]]


In [4]:
BI2 = RowAdd(BI1,0,2,2)
print(BI2)

[[0. 1. 0. 1. 1. 0. 0. 0.]
 [0. 1. 0. 6. 4. 1. 0. 0.]
 [1. 1. 3. 3. 2. 0. 1. 0.]
 [2. 4. 0. 5. 0. 0. 0. 1.]]


In [5]:
BI3 = RowAdd(BI2,0,3,-4)
print(BI3)

[[ 0.  1.  0.  1.  1.  0.  0.  0.]
 [ 0.  1.  0.  6.  4.  1.  0.  0.]
 [ 1.  1.  3.  3.  2.  0.  1.  0.]
 [ 2.  0.  0.  1. -4.  0.  0.  1.]]


In [6]:
BI4 = RowScale(BI3,2,2)
print(BI4)

[[ 0.  1.  0.  1.  1.  0.  0.  0.]
 [ 0.  1.  0.  6.  4.  1.  0.  0.]
 [ 2.  2.  6.  6.  4.  0.  2.  0.]
 [ 2.  0.  0.  1. -4.  0.  0.  1.]]


In [7]:
BI5 = RowScale(BI4,2,-1)
print(BI5)

[[ 0.  1.  0.  1.  1.  0.  0.  0.]
 [ 0.  1.  0.  6.  4.  1.  0.  0.]
 [-2. -2. -6. -6. -4. -0. -2. -0.]
 [ 2.  0.  0.  1. -4.  0.  0.  1.]]


In [8]:
BI6 = RowScale(BI5,3,-1)
print(BI6)

[[ 0.  1.  0.  1.  1.  0.  0.  0.]
 [ 0.  1.  0.  6.  4.  1.  0.  0.]
 [-2. -2. -6. -6. -4. -0. -2. -0.]
 [-2. -0. -0. -1.  4. -0. -0. -1.]]


In [9]:
BI7= RowAdd(BI6,1,0,-1)
print(BI7)

[[ 0.  0.  0. -5. -3. -1.  0.  0.]
 [ 0.  1.  0.  6.  4.  1.  0.  0.]
 [-2. -2. -6. -6. -4. -0. -2. -0.]
 [-2. -0. -0. -1.  4. -0. -0. -1.]]


In [10]:
BI8 = RowAdd(BI7,1,2,-1)
print(BI8)

[[  0.   0.   0.  -5.  -3.  -1.   0.   0.]
 [  0.   1.   0.   6.   4.   1.   0.   0.]
 [ -2.  -3.  -6. -12.  -8.  -1.  -2.  -0.]
 [ -2.  -0.  -0.  -1.   4.  -0.  -0.  -1.]]


In [11]:
BI9 = RowAdd(BI8,1,3,-1)
print(BI9)

[[  0.   0.   0.  -5.  -3.  -1.   0.   0.]
 [  0.   1.   0.   6.   4.   1.   0.   0.]
 [ -2.  -3.  -6. -12.  -8.  -1.  -2.  -0.]
 [ -2.  -1.  -0.  -7.   0.  -1.  -0.  -1.]]


In [12]:
BI10 = RowAdd(BI9,3,0,-3)
print(BI10)

[[  6.   3.   0.  16.  -3.   2.   0.   3.]
 [  0.   1.   0.   6.   4.   1.   0.   0.]
 [ -2.  -3.  -6. -12.  -8.  -1.  -2.  -0.]
 [ -2.  -1.  -0.  -7.   0.  -1.  -0.  -1.]]


In [13]:
BI11 = RowAdd(BI10,3,1,5)
print(BI11)

[[  6.   3.   0.  16.  -3.   2.   0.   3.]
 [-10.  -4.   0. -29.   4.  -4.   0.  -5.]
 [ -2.  -3.  -6. -12.  -8.  -1.  -2.  -0.]
 [ -2.  -1.  -0.  -7.   0.  -1.  -0.  -1.]]


In [14]:
BI12 = RowAdd(BI11,3,2,-2)
print(BI12)

[[  6.   3.   0.  16.  -3.   2.   0.   3.]
 [-10.  -4.   0. -29.   4.  -4.   0.  -5.]
 [  2.  -1.  -6.   2.  -8.   1.  -2.   2.]
 [ -2.  -1.  -0.  -7.   0.  -1.  -0.  -1.]]


In [15]:
BI13 = RowScale(BI12,2,-1/11)
print(BI13)

[[  6.           3.           0.          16.          -3.
    2.           0.           3.        ]
 [-10.          -4.           0.         -29.           4.
   -4.           0.          -5.        ]
 [ -0.18181818   0.09090909   0.54545455  -0.18181818   0.72727273
   -0.09090909   0.18181818  -0.18181818]
 [ -2.          -1.          -0.          -7.           0.
   -1.          -0.          -1.        ]]


In [16]:
BI14 =  RowAdd(BI13,2,0,28)
print(BI14)

[[  0.90909091   5.54545455  15.27272727  10.90909091  17.36363636
   -0.54545455   5.09090909  -2.09090909]
 [-10.          -4.           0.         -29.           4.
   -4.           0.          -5.        ]
 [ -0.18181818   0.09090909   0.54545455  -0.18181818   0.72727273
   -0.09090909   0.18181818  -0.18181818]
 [ -2.          -1.          -0.          -7.           0.
   -1.          -0.          -1.        ]]


In [17]:
BI15 = RowAdd(BI14,2,1,-49)
print(BI15)

[[  0.90909091   5.54545455  15.27272727  10.90909091  17.36363636
   -0.54545455   5.09090909  -2.09090909]
 [ -1.09090909  -8.45454545 -26.72727273 -20.09090909 -31.63636364
    0.45454545  -8.90909091   3.90909091]
 [ -0.18181818   0.09090909   0.54545455  -0.18181818   0.72727273
   -0.09090909   0.18181818  -0.18181818]
 [ -2.          -1.          -0.          -7.           0.
   -1.          -0.          -1.        ]]


In [18]:
BI16 = RowAdd(BI15,2,3,-9)
print(BI16)

[[  0.90909091   5.54545455  15.27272727  10.90909091  17.36363636
   -0.54545455   5.09090909  -2.09090909]
 [ -1.09090909  -8.45454545 -26.72727273 -20.09090909 -31.63636364
    0.45454545  -8.90909091   3.90909091]
 [ -0.18181818   0.09090909   0.54545455  -0.18181818   0.72727273
   -0.09090909   0.18181818  -0.18181818]
 [ -0.36363636  -1.81818182  -4.90909091  -5.36363636  -6.54545455
   -0.18181818  -1.63636364   0.63636364]]


In [19]:
BI17 = RowScale(BI16,3,-1)
print(BI17)

[[  0.90909091   5.54545455  15.27272727  10.90909091  17.36363636
   -0.54545455   5.09090909  -2.09090909]
 [ -1.09090909  -8.45454545 -26.72727273 -20.09090909 -31.63636364
    0.45454545  -8.90909091   3.90909091]
 [ -0.18181818   0.09090909   0.54545455  -0.18181818   0.72727273
   -0.09090909   0.18181818  -0.18181818]
 [  0.36363636   1.81818182   4.90909091   5.36363636   6.54545455
    0.18181818   1.63636364  -0.63636364]]


In [20]:
BI18 = RowSwap(BI17,3,0)
print(BI18)

[[  0.36363636   1.81818182   4.90909091   5.36363636   6.54545455
    0.18181818   1.63636364  -0.63636364]
 [ -1.09090909  -8.45454545 -26.72727273 -20.09090909 -31.63636364
    0.45454545  -8.90909091   3.90909091]
 [ -0.18181818   0.09090909   0.54545455  -0.18181818   0.72727273
   -0.09090909   0.18181818  -0.18181818]
 [  0.90909091   5.54545455  15.27272727  10.90909091  17.36363636
   -0.54545455   5.09090909  -2.09090909]]


In [21]:
BI19 = RowSwap(BI18,3,1)
print(BI19)

[[  0.36363636   1.81818182   4.90909091   5.36363636   6.54545455
    0.18181818   1.63636364  -0.63636364]
 [  0.90909091   5.54545455  15.27272727  10.90909091  17.36363636
   -0.54545455   5.09090909  -2.09090909]
 [ -0.18181818   0.09090909   0.54545455  -0.18181818   0.72727273
   -0.09090909   0.18181818  -0.18181818]
 [ -1.09090909  -8.45454545 -26.72727273 -20.09090909 -31.63636364
    0.45454545  -8.90909091   3.90909091]]


In [22]:
BI20 = RowSwap(BI19,2,3)
print(BI20)

[[  0.36363636   1.81818182   4.90909091   5.36363636   6.54545455
    0.18181818   1.63636364  -0.63636364]
 [  0.90909091   5.54545455  15.27272727  10.90909091  17.36363636
   -0.54545455   5.09090909  -2.09090909]
 [ -1.09090909  -8.45454545 -26.72727273 -20.09090909 -31.63636364
    0.45454545  -8.90909091   3.90909091]
 [ -0.18181818   0.09090909   0.54545455  -0.18181818   0.72727273
   -0.09090909   0.18181818  -0.18181818]]


In [23]:
B = np.array([[0,1,0,1],[0,-3,0,2],[1,-1,3,1],[2,4,0,5]])
BS = np.linalg.inv(B)
print(BS)

[[-2.3        -0.1         0.          0.5       ]
 [ 0.4        -0.2        -0.         -0.        ]
 [ 0.7        -0.1         0.33333333 -0.16666667]
 [ 0.6         0.2         0.          0.        ]]


Setelah ketemu invers matriksnya kita operasikan untuk menemukan X dengan mengalikan matriks A-1
 dengan matriks B

$$
\begin{align} \begin{bmatrix} 1&0&0&0\\ 0&1&0&0\\ 0&0&1&0\\ 0&0&0&1\\  \end{bmatrix} \begin{bmatrix} x_1\\ x_2\\ x_3\\ x_4\\ \end{bmatrix}  - \end{align}
$$

$$
\begin{align} \begin{bmatrix} 0.45454545&-0.18181818&0.81818182&-0.63636364 \\ -0.36363636&0.54545455&2.54545455&-2.09090909 \\-0.36363636&-0.45454545&-4.45454545&3.90909091 \\ 0.27272727&0.09090909&0.09090909&-0.18181818\\  \end{bmatrix} \begin{bmatrix} 6\\ 10\\ 11\\ 14\\ \end{bmatrix} \end{align}
$$

Setelah itu kita operasikan menggunakan fungsi NumPy pada Python

In [24]:
B = np.array ([[6,10,11,14]])
BT = B.transpose()
X = np.dot(BS,BT)
print(X)

[[-7.8       ]
 [ 0.4       ]
 [ 4.53333333]
 [ 5.6       ]]


Dari operasi diatas bisa kita dapatkan nilai dari X1,X2,X3,X4

$$
x_1 = 1x_2 = 2x_3 = -1x_4 = 1
$$

References:
* [https://kumparan.com/berita-terkini/contoh-soal-eliminasi-gauss-lengkap-dengan-pembahasannya-1ynbQVj8e5w]
* [https://yuliantidewi2018.blogspot.com/2018/04/eliminasi-gauss.html]
